File: lua/distract/entity_step.lua
--- One entity, one simulated frame.
---
--- Extracted from `engine.lua`'s `M.step`, which had grown past 200 lines and was
--- the reason that module sat over a 400-line cap. Verbatim: the same arithmetic in
--- the same order, and the physics-parity fixtures are the characterization tests
--- that say so -- a golden that moved would mean the move changed behaviour.
---
--- Everything the frame needs and this module cannot see is passed in rather than
--- reached for: the bounds, the obstacle list, the collision sink, and the two
--- engine-owned callbacks -- `set_state`, which has to run the engine's own
--- transition bookkeeping, and `sprite_cell_size`, which resolves an asset's
--- footprint through the sprite cache.
---
--- `M.advance` is one long function on purpose. Its five numbered steps look like
--- five functions and are not separable: they share a dozen locals -- the
--- footprint, the resolved physics, the per-axis cell scales -- and each step reads
--- what the previous one wrote, so splitting them means threading that state
--- through five signatures or rebuilding it five times. The numbered comments are
--- the structure. `../../CLAUDE.md` §5 covers this case: a cap is a signal to
--- decompose, not a reason to fragment a unit that has to be read as one.
---
--- Units are terminal cells, as everywhere else in this engine.

local M = {}

local kinematics = require("distract.kinematics")
local locomotion = require("distract.locomotion")
local obstacles = require("distract.obstacles")

local CELLS_PER_SPRITE_PX_X = kinematics.CELLS_PER_SPRITE_PX_X
local CELLS_PER_SPRITE_PX_Y = kinematics.CELLS_PER_SPRITE_PX_Y
local BALLISTIC = locomotion.BALLISTIC

---@class DistractFrame
---@field dt number seconds this frame covers
---@field step number `dt` in frames at the reference 60 FPS
---@field min_col number left edge of the bounds, in cells
---@field min_row number top edge
---@field max_columns number right edge
---@field max_lines number bottom edge
---@field obstacle_rects table[] platforms and hazards, in cells
---@field collisions table[] appended to, drained by the caller
---@field set_state function `(entity, state)`
---@field sprite_cell_size function `(asset_name) -> width, height`

--- Advances one entity, reporting whether it left the world.
---@param entity table
---@param frame DistractFrame
---@return boolean whether the entity deactivated itself and needs collecting
function M.advance(entity, frame)
  local dt, step = frame.dt, frame.step
  local min_col, min_row = frame.min_col, frame.min_row
  local max_columns, max_lines = frame.max_columns, frame.max_lines
  local obstacle_rects = frame.obstacle_rects
  local collisions = frame.collisions
  local set_state = frame.set_state
  local sprite_cell_size = frame.sprite_cell_size
  local despawned = false

  entity.state_time = entity.state_time + dt

  -- 1. Action duration timer
  if entity.action_timer and entity.action_duration then
    entity.action_timer = entity.action_timer + dt
    if entity.action_timer >= entity.action_duration then
      entity.action_timer = nil
      entity.action_duration = nil
      entity.is_locked = false
      local next_state = entity.return_state or "idle"
      entity.return_state = nil
      set_state(entity, next_state)
    end
  end

  local state_def = entity.manifest.states and entity.manifest.states[entity.current_state]
  if state_def then
    -- 2. State Timeout
    if
      state_def.transitions
      and state_def.transitions.timeout_ms
      and state_def.transitions.on_timeout
    then
      if entity.state_time * 1000 >= state_def.transitions.timeout_ms then
        set_state(entity, state_def.transitions.on_timeout)
      end
    end

    -- 3. Animation frames
    local anim = state_def.animation or { frames = { 0 }, fps = 6, loop_anim = true }
    local frame_count = #(anim.frames or { 0 })
    if frame_count > 0 then
      local frame_duration = kinematics.frame_duration_seconds(entity, anim)
      entity.frame_timer = entity.frame_timer + dt

      if entity.frame_timer >= frame_duration then
        entity.frame_timer = entity.frame_timer - frame_duration
        if entity.frame_idx < frame_count then
          entity.frame_idx = entity.frame_idx + 1
        elseif anim.loop_anim ~= false then
          entity.frame_idx = 1
        else
          entity.animation_finished = true
          if state_def.transitions and state_def.transitions.on_finish then
            set_state(entity, state_def.transitions.on_finish)
          end
        end
      end
    end

    -- 4. Physics, in the shared manifest unit (sprite pixels per 60 FPS frame)
    --
    -- Parallax damps the displacement rather than the velocity itself:
    -- damping the stored velocity every frame would decay it to zero instead
    -- of moving a distant thing slower at a steady speed.
    local parallax = entity.parallax or 1.0
    local cells_x = step * CELLS_PER_SPRITE_PX_X * parallax
    local cells_y = step * CELLS_PER_SPRITE_PX_Y * parallax
    -- The footprint every surface and boundary is measured against. Sizes come
    -- from the asset rather than a constant: the built-in sprites are 24 cells
    -- wide (cat, crab) and 16 (sun), so a hardcoded 16 measured in the wrong
    -- place. Parallax shrinks the drawn art, so the footprint shrinks with it.
    local sprite_w, sprite_h = sprite_cell_size(entity.asset_name)
    sprite_w, sprite_h = sprite_w * parallax, sprite_h * parallax
    local phys = state_def.physics or {}
    local speed_x = math.abs(phys.target_vx or 0)
    entity.target_vx = speed_x * entity.heading_x
    entity.target_vy = phys.target_vy or 0
    entity.flip_x = (entity.heading_x < 0)

    local friction = phys.friction or 0.05
    local lerp_factor = math.min(1.0, math.max(0.01, 1.0 - math.exp(-friction * step)))
    entity.vx = entity.vx + (entity.target_vx - entity.vx) * lerp_factor
    -- Constant acceleration, on top of the pull toward `target_vx`. `gravity`
    -- is the same thing on the y axis under a name that also brings a floor
    -- with it; `accel_y` is the floorless version.
    entity.vx = entity.vx + ((phys.accel_x or 0) * step)

    if (phys.gravity or 0) > 0 then
      -- Read before the integration: an entity already resting on the floor
      -- is re-accelerated by gravity and caught by the clamp on every single
      -- tick, so "the clamp ran" is not a landing. Crossing the floor from
      -- above is.
      local feet_before = entity.y + sprite_h
      entity.vy = entity.vy + (phys.gravity * step)
      entity.y = entity.y + (entity.vy * cells_y)

      -- A registered platform is a floor the entity reaches earlier, so the
      -- surface for this frame is whichever is higher. With no obstacles this is
      -- the floor exactly, and the arithmetic is the ground clamp it replaces.
      local floor_feet = entity.ground_y + sprite_h
      local surface = floor_feet
      local platform_top = obstacles.crossed_platform(obstacle_rects, {
        left = entity.x,
        top = entity.y,
        width = sprite_w,
        height = sprite_h,
      }, feet_before)
      if platform_top and platform_top < surface then
        surface = platform_top
      end
      local was_airborne = feet_before < surface

      if entity.y + sprite_h >= surface then
        entity.y = surface - sprite_h
        local landed = was_airborne and entity.vy > 0
        entity.vy = 0
        if landed then
          table.insert(collisions, { entity = entity, edge = "bottom" })
        end
        if landed and locomotion.effective_locomotion(phys) == BALLISTIC then
          local on_land = state_def.transitions and state_def.transitions.on_land
          if on_land then
            -- Landing ends the action that launched the entity. Leaving its
            -- timer running would drag the entity out of the state it just
            -- reached as soon as the clock caught up, so a jump that lands
            -- early would still be locked until its declared duration.
            entity.action_timer = nil
            entity.action_duration = nil
            entity.return_state = nil
            entity.is_locked = false
            set_state(entity, on_land)
          end
        end
      end
    else
      entity.vy = entity.vy + (entity.target_vy - entity.vy) * lerp_factor
      entity.vy = entity.vy + ((phys.accel_y or 0) * step)
      entity.y = entity.y + (entity.vy * cells_y)
    end

    entity.x = entity.x + (entity.vx * cells_x)

    -- A path is a positional *override*, applied after integration so it
    -- replaces the velocity result on the axes it owns and leaves the others
    -- alone. Gravity is excluded: a path that writes y fights the floor,
    -- which is what the locomotion classes exist to keep apart.
    if (phys.gravity or 0) <= 0 then
      kinematics.apply_path(entity, phys, dt)
    end

    -- A grounded state has no gravity to fall under, so which surface it stands
    -- on is resolved rather than integrated. Only reached while obstacles exist:
    -- without them the answer is the floor it was seated on.
    if
      #obstacle_rects > 0
      and (phys.gravity or 0) <= 0
      and entity.ground_y
      and locomotion.locomotion_for(entity.manifest, state_def) == locomotion.GROUNDED
    then
      entity.y = obstacles.standing_surface(obstacle_rects, {
        left = entity.x,
        top = entity.y,
        width = sprite_w,
        height = sprite_h,
      }, entity.ground_y + sprite_h) - sprite_h
    end

    local deflection = obstacles.deflection(obstacle_rects, {
      left = entity.x,
      top = entity.y,
      width = sprite_w,
      height = sprite_h,
    }, entity.heading_x)
    if deflection then
      entity.x = deflection.x
      entity.heading_x = deflection.heading_x
      entity.vx = math.abs(entity.vx) * entity.heading_x
      entity.flip_x = entity.heading_x < 0
      table.insert(collisions, { entity = entity, edge = "obstacle" })
    end

    -- 5. Screen boundary modes.
    local wrap_mode = phys.wrap_mode or "wrap"

    local edges = state_def.transitions or {}

    if wrap_mode == "wrap" then
      -- Gated on position, not on velocity: `vx` lerps toward its target, so
      -- a state whose target is zero decays it through zero and an entity
      -- that had already left the screen would never wrap back.
      if entity.x > max_columns then
        entity.x = min_col - sprite_w
      elseif entity.x < min_col - sprite_w then
        entity.x = max_columns
      end
      -- Vertical wrap too. The overlay has always wrapped both axes, so a
      -- manifest with vertical motion described one behaviour there and a
      -- different one here.
      if entity.y > max_lines then
        entity.y = min_row - sprite_h
      elseif entity.y < min_row - sprite_h then
        entity.y = max_lines
      end
    elseif wrap_mode == "bounce" then
      if entity.x <= min_col then
        entity.x = min_col
        entity.heading_x = 1
        entity.vx = math.max(0.5, math.abs(entity.vx))
        entity.flip_x = false
        table.insert(collisions, { entity = entity, edge = "left" })
        if edges.on_edge_left then
          set_state(entity, edges.on_edge_left)
        end
      elseif entity.x + sprite_w >= max_columns then
        entity.x = math.max(min_col, max_columns - sprite_w)
        entity.heading_x = -1
        entity.vx = -math.max(0.5, math.abs(entity.vx))
        entity.flip_x = true
        table.insert(collisions, { entity = entity, edge = "right" })
        if edges.on_edge_right then
          set_state(entity, edges.on_edge_right)
        end
      end

      if entity.vy ~= 0 then
        if entity.y <= min_row then
          entity.y = min_row
          entity.vy = math.abs(entity.vy)
          table.insert(collisions, { entity = entity, edge = "top" })
        elseif entity.y + sprite_h >= max_lines then
          entity.y = math.max(min_row, max_lines - sprite_h)
          entity.vy = -math.abs(entity.vy)
          table.insert(collisions, { entity = entity, edge = "bottom" })
        end
      end
    elseif wrap_mode == "clamp" then
      entity.x = math.max(min_col, math.min(entity.x, max_columns - sprite_w))
      -- No `- 1` on the ceiling: the overlay clamps at `viewport_h - frame_h`
      -- and `bounce` above clamps at `max_lines - sprite_h`, so the stray row
      -- made `clamp` disagree with both the other engine and its own
      -- neighbouring branch. Reserving space for the statusline is the floor
      -- system's job, where it is computed from `cmdheight` and `laststatus`
      -- rather than guessed at as a constant.
      entity.y = math.max(min_row, math.min(entity.y, max_lines - sprite_h))
    elseif wrap_mode == "despawn" then
      if
        entity.x < min_col - sprite_w
        or entity.x > max_columns
        or entity.y < min_row - sprite_h
        or entity.y > max_lines
      then
        entity.is_active = false
        despawned = true
      end
    end
    -- "none" deliberately applies no boundary handling.
  end

  return despawned
end

return M

File: engine/src/entity_step.rs
//! One entity's frame: timers, animation, physics, then boundaries.
//!
//! The Rust counterpart to `lua/distract/entity_step.lua`, and split from
//! `ecs.rs` for the same reason that one was split from `engine.lua`:
//! `World::update` was three hundred lines of per-entity work wrapped in thirty
//! lines of coordination, and the two read as one only because they shared a
//! scope.
//!
//! The five numbered steps run in order and each reads what the previous one
//! wrote, which is why this is one function rather than five. The physics-parity
//! fixtures pin that order across both engines.

use crate::asset::AssetManager;
use crate::bounds::Bounds;
use crate::entity::Entity;
use crate::frame_timing::frame_duration_seconds;
use crate::journal::{self, WorldEvent};
use crate::manifest;
use crate::manifest::WrapMode;
use crate::obstacles::{self, Footprint, Obstacle, PushDirection};
use crate::path::apply_path;

/// Everything one entity's step reads that is not the entity.
///
/// Assembled once per frame by `World::update`: the entity list is borrowed
/// mutably for the whole loop, so nothing here can be reached through `self`.
pub struct StepContext<'a> {
    pub dt: f32,
    pub bounds: Bounds,
    /// Terminal cell size, which is what converts a manifest's sprite pixels
    /// into the pixels the boundary modes measure against.
    pub scale_x: f32,
    pub scale_y: f32,
    pub assets: &'a AssetManager,
    pub obstacles: &'a [Obstacle],
    /// Whether the journal is recording, so collision events are only built when
    /// something will read them.
    pub is_recording: bool,
}

/// Advances one entity by `context.dt`, appending any collision it reports.
pub fn advance(entity: &mut Entity, context: &StepContext, collisions: &mut Vec<WorldEvent>) {
    if !entity.is_active {
        return;
    }

    entity.state_time += context.dt;

    // 1. Advance action timer and handle return state
    if let (Some(ref mut timer), Some(duration)) =
        (&mut entity.action_timer, entity.action_duration)
    {
        *timer += context.dt;
        if *timer >= duration {
            entity.action_timer = None;
            entity.action_duration = None;
            entity.is_locked = false;
            let next_state = entity
                .return_state
                .take()
                .unwrap_or_else(|| "idle".to_string());
            entity.set_state(next_state);
        }
    }

    let asset = match context.assets.get(&entity.asset_name) {
        Some(a) => a,
        None => return,
    };

    // Parallax shrinks the drawn art, so the footprint the boundary
    // modes measure against shrinks with it.
    let frame_w = asset.frame_w as f32 * context.scale_x * entity.parallax;
    let frame_h = asset.frame_h as f32 * context.scale_y * entity.parallax;

    let state_def = asset.manifest.states.get(&entity.current_state);

    if let Some(state_def) = state_def {
        // 2. Check timeout transitions
        if let (Some(timeout_ms), Some(ref next_state)) = (
            state_def.transitions.timeout_ms,
            &state_def.transitions.on_timeout,
        ) {
            if entity.state_time * 1000.0 >= timeout_ms as f32 {
                entity.set_state(next_state.clone());
            }
        }

        // 3. Advance animation frames
        let anim = &state_def.animation;
        let frame_count = anim.frames.len();
        if frame_count > 0 {
            let frame_duration = frame_duration_seconds(anim, entity.frame_idx, asset);
            entity.frame_timer += context.dt;

            if entity.frame_timer >= frame_duration {
                entity.frame_timer -= frame_duration;
                if entity.frame_idx + 1 < frame_count {
                    entity.frame_idx += 1;
                } else if anim.loop_anim {
                    entity.frame_idx = 0;
                } else {
                    entity.animation_finished = true;
                    if let Some(ref next_state) = state_def.transitions.on_finish {
                        entity.set_state(next_state.clone());
                    }
                }
            }
        }

        // 4. Physics, in the shared manifest unit.
        //
        // Manifest positions and velocities are in *sprite pixels* per
        // frame at 60 FPS, and one sprite pixel is one terminal cell
        // wide. Converting on integration is what makes one manifest
        // describe one behaviour on both backends; the two used to
        // apply unrelated ad-hoc factors and moved at different speeds.
        //
        // Parallax damps the displacement rather than the stored
        // velocity: damping the velocity every frame would decay it to
        // zero instead of moving a distant thing slower at a steady
        // speed.
        let step = context.dt * 60.0;
        let px = step * context.scale_x * entity.parallax;
        let py = step * context.scale_y * entity.parallax;
        let phys = &state_def.physics;
        let speed_x = phys.target_vx.abs();
        entity.target_vx = speed_x * entity.heading_x;
        entity.target_vy = phys.target_vy;

        // Sync flip with heading direction
        entity.flip_x = entity.heading_x < 0.0;

        // Smooth exponential velocity lerping
        let lerp_factor = (1.0 - (-phys.friction * step).exp()).clamp(0.01, 1.0);
        entity.vx += (entity.target_vx - entity.vx) * lerp_factor;
        // Constant acceleration, on top of the pull toward `target_vx`.
        // These were declared in the manifest schema and read by
        // nothing, so a manifest could set them and watch them do
        // nothing. `gravity` is `accel_y` under a name that also brings
        // a floor with it.
        entity.vx += phys.accel_x * step;

        if phys.gravity > 0.0 {
            // Read before the integration: an entity already resting on
            // the floor is re-accelerated by gravity and caught by the
            // clamp on every single tick, so "the clamp ran" is not a
            // landing. Crossing the floor from above is.
            let feet_before = entity.y + frame_h;
            entity.vy += phys.gravity * step;
            entity.y += entity.vy * py;

            // A registered platform is a floor the entity reaches
            // earlier, so the surface for this frame is whichever is
            // higher. With no context.obstacles this is the floor exactly, and
            // the arithmetic is the ground clamp it replaces.
            let floor_feet = entity.ground_y + frame_h;
            let surface = obstacles::crossed_platform(
                context.obstacles,
                Footprint {
                    left: entity.x,
                    top: entity.y,
                    width: frame_w,
                    height: frame_h,
                },
                feet_before,
            )
            .map_or(floor_feet, |platform_top| platform_top.min(floor_feet));
            let was_airborne = feet_before < surface;

            if entity.y + frame_h >= surface {
                entity.y = surface - frame_h;
                let landed = was_airborne && entity.vy > 0.0;
                entity.vy = 0.0;
                if landed && context.is_recording {
                    collisions.push(WorldEvent::collision(entity.id, journal::EDGE_BOTTOM));
                }
                if landed && phys.effective_locomotion() == manifest::BALLISTIC {
                    if let Some(ref land_state) = state_def.transitions.on_land {
                        // Landing ends the action that launched the
                        // entity. Leaving its timer running would drag
                        // the entity out of the state it just reached
                        // as soon as the clock caught up, so a jump
                        // that lands early would still be locked until
                        // its declared duration.
                        entity.action_timer = None;
                        entity.action_duration = None;
                        entity.return_state = None;
                        entity.is_locked = false;
                        entity.set_state(land_state.clone());
                    }
                }
            }
        } else {
            entity.vy += (entity.target_vy - entity.vy) * lerp_factor;
            entity.vy += phys.accel_y * step;
            entity.y += entity.vy * py;
        }

        entity.x += entity.vx * px;

        // A path is a positional *override*, applied after integration
        // so it replaces the velocity result on the axes it owns and
        // leaves the others alone. Gravity is excluded: a path that
        // writes y fights the floor, which is what the locomotion
        // classes exist to keep apart.
        if phys.gravity <= 0.0 {
            if let Some(ref path_type) = phys.path_type {
                apply_path(
                    entity,
                    path_type,
                    phys,
                    context.dt,
                    context.scale_x,
                    context.scale_y,
                );
            }
        }

        // A grounded state has no gravity to fall under, so which
        // surface it stands on is resolved rather than integrated. Only
        // reached while context.obstacles exist: without them the answer is the
        // floor the entity was already seated on.
        if !context.obstacles.is_empty()
            && phys.gravity <= 0.0
            && asset.manifest.locomotion_for(state_def) == manifest::GROUNDED
        {
            entity.y = obstacles::standing_surface(
                context.obstacles,
                Footprint {
                    left: entity.x,
                    top: entity.y,
                    width: frame_w,
                    height: frame_h,
                },
                entity.ground_y + frame_h,
            ) - frame_h;
        }

        if let Some(deflection) = obstacles::deflection(
            context.obstacles,
            Footprint {
                left: entity.x,
                top: entity.y,
                width: frame_w,
                height: frame_h,
            },
            entity.heading_x,
        ) {
            entity.x = deflection.x;
            entity.heading_x = match deflection.direction {
                PushDirection::Left => -1.0,
                PushDirection::Right => 1.0,
            };
            entity.vx = entity.vx.abs() * entity.heading_x;
            entity.flip_x = entity.heading_x < 0.0;
            if context.is_recording {
                collisions.push(WorldEvent::collision(entity.id, journal::EDGE_OBSTACLE));
            }
        }

        // 5. Boundary checking
        match phys.wrap_mode {
            WrapMode::Wrap => {
                // Gated on position and heading, not on instantaneous
                // velocity: `vx` lerps toward its target, so a state
                // whose target is zero decays it through zero and an
                // entity that had already left the viewport would never
                // wrap back — it just sat off-screen forever.
                if entity.x > context.bounds.right() {
                    entity.x = context.bounds.left - frame_w;
                } else if entity.x < context.bounds.left - frame_w {
                    entity.x = context.bounds.right();
                }
                if entity.y > context.bounds.bottom() {
                    entity.y = context.bounds.top - frame_h;
                } else if entity.y < context.bounds.top - frame_h {
                    entity.y = context.bounds.bottom();
                }
            }
            WrapMode::Bounce => {
                if entity.x <= context.bounds.left {
                    entity.x = context.bounds.left;
                    entity.heading_x = 1.0;
                    entity.vx = entity.vx.abs().max(0.5);
                    entity.flip_x = false;
                    if context.is_recording {
                        collisions.push(WorldEvent::collision(entity.id, journal::EDGE_LEFT));
                    }
                    if let Some(ref edge_state) = state_def.transitions.on_edge_left {
                        entity.set_state(edge_state.clone());
                    }
                } else if entity.x + frame_w >= context.bounds.right() {
                    entity.x = (context.bounds.right() - frame_w).max(context.bounds.left);
                    entity.heading_x = -1.0;
                    entity.vx = -entity.vx.abs().max(0.5);
                    entity.flip_x = true;
                    if context.is_recording {
                        collisions.push(WorldEvent::collision(entity.id, journal::EDGE_RIGHT));
                    }
                    if let Some(ref edge_state) = state_def.transitions.on_edge_right {
                        entity.set_state(edge_state.clone());
                    }
                }

                if entity.vy != 0.0 {
                    if entity.y <= context.bounds.top {
                        entity.y = context.bounds.top;
                        entity.vy = entity.vy.abs();
                        if context.is_recording {
                            collisions.push(WorldEvent::collision(entity.id, journal::EDGE_TOP));
                        }
                    } else if entity.y + frame_h >= context.bounds.bottom() {
                        entity.y = (context.bounds.bottom() - frame_h).max(context.bounds.top);
                        entity.vy = -entity.vy.abs();
                        if context.is_recording {
                            collisions.push(WorldEvent::collision(entity.id, journal::EDGE_BOTTOM));
                        }
                    }
                }
            }
            WrapMode::Clamp => {
                entity.x = entity.x.clamp(
                    context.bounds.left,
                    (context.bounds.right() - frame_w).max(context.bounds.left),
                );
                entity.y = entity.y.clamp(
                    context.bounds.top,
                    (context.bounds.bottom() - frame_h).max(context.bounds.top),
                );
            }
            WrapMode::Despawn => {
                if entity.x < context.bounds.left - frame_w
                    || entity.x > context.bounds.right()
                    || entity.y < context.bounds.top - frame_h
                    || entity.y > context.bounds.bottom()
                {
                    entity.is_active = false;
                }
            }
            WrapMode::None => {}
        }
    }
}
